modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SSource.java

```diff
@@ -186,6 +185,20 @@ void analyze(Locals program) {
 
         mainMethod = Locals.newMainMethodScope(scriptInterface, program, reserved.getMaxLoopCounter());
 
+        for (int get = 0; get < scriptInterface.getGetMethods().size();) {
+            String name = scriptInterface.getGetMethods().get(get).getName().substring(3);
+            name = Character.toLowerCase(name.charAt(0)) + name.substring(1);
+            Definition.Type rtn = scriptInterface.getGetReturns().get(get);
+
+            if (reserved.getUsedVariables().contains(name)) {
+                mainMethod.addVariable(new Location("getter [" + name + "]", 0), rtn, name, true);
+                ++get;
+            } else {
+                scriptInterface.getGetMethods().remove(get);
```

Comment:
This seems quite sneaky that you modify the lists returned by `getGetMethods()` and `getGetReturns()`. Additionally, I'm not sure how it actually works since you increment `get`, but then but remove, so the size is changing.

Can you instead create a copy of the lists in SSource ctor and use those local versions? I think you could then just create the lists on SSource only containing variables contained in `reserved.getUsedVariables()`? Then just iterate the member lists here to add to the `mainMethod`.